File: src/structural/decorator2.ts
class InitialDatabaseService {
  getMailFromUsername(username: string): string {
    return username + "@mail";
  }
}

class DatabaseService extends InitialDatabaseService {
  getPhoneNbrFromUsername(username: string): string {
    return username + "@Phone";
  }

  getFBNameFromUsername(username: string): string {
    return username + "@Facebook";
  }
}

interface INotifier {
  send(message: string): void;
  getUsername(): string;
}

class Notifier implements INotifier {
  private databaseService = new DatabaseService();

  constructor(private username: string) {}

  send(msg: string): void {
    const mail = this.databaseService.getMailFromUsername(this.username);
    console.log("Sending " + msg + " by Mail to " + mail);
  }

  getUsername(): string {
    return this.username;
  }
}

abstract class BaseNotifierDecorator implements INotifier {
  protected readonly databaseService = new DatabaseService();

  constructor(private wrapped: INotifier) {}

  send(msg: string): void {
    this.wrapped.send(msg);
  }

  getUsername(): string {
    return this.wrapped.getUsername();
  }
}

class FacebookDecorator extends BaseNotifierDecorator {
  constructor(wrapped: INotifier) {
    super(wrapped);
  }

  override send(msg: string): void {
    super.send(msg);
    const fbName = this.databaseService.getFBNameFromUsername(this.getUsername());
    console.log("Sending " + msg + " on Facebook to " + fbName);
  }
}

class WhatsAppDecorator extends BaseNotifierDecorator {
  constructor(wrapped: INotifier) {
    super(wrapped);
  }

  override send(msg: string): void {
    super.send(msg);
    const phoneNbr = this.databaseService.getPhoneNbrFromUsername(this.getUsername());
    console.log("Sending " + msg + " by WhatsApp on " + phoneNbr);
  }
}

function sendNotifications(): void {
  const notifier = new FacebookDecorator(new WhatsAppDecorator(new Notifier("Geekific")));
  notifier.send("Like and Subscribe!!!");
}

sendNotifications();

File: src/structural/decorator2.java
public interface INotifier {
    void send(String message);
    String getUsername();
}

public class Notifier implements INotifier {
    private final String username;
    private final DatabaseService databaseService;

    public Notifier(String username) {
        this.username = username;
        databaseService = new DatabaseService();
    }

    @Override
    public void send(String msg) {
        String mail = databaseService.getMailFromUsername(username);
        System.out.println("Sending " + msg + " by Mail to " + mail);
    }

    @Override
    public String getUsername() {
        return username;
    }
}

public abstract class BaseNotifierDecorator implements INotifier {
    private final INotifier wrapped;
    protected final DatabaseService databaseService;

    public BaseNotifierDecorator(INotifier wrapped) {
        this.wrapped = wrapped;
        databaseService = new DatabaseService();
    }

    @Override
    public void send(String msg) {
        wrapped.send(msg);
    }

    @Override
    public String getUsername() {
        return wrapped.getUsername();
    }
}

public class FacebookDecorator extends BaseNotifierDecorator {
    public FacebookDecorator(INotifier wrapped) {
        super(wrapped);
    }

    @Override
    public void send(String msg) {
        super.send(msg);
        String fbName = databaseService.getFBNameFromUsername(getUsername());
        System.out.println("Sending " + msg + " on Facebook to " + fbName);
    }
}

public class WhatsAppDecorator extends BaseNotifierDecorator {
    public WhatsAppDecorator(INotifier wrapped) {
        super(wrapped);
    }

    @Override
    public void send(String msg) {
        super.send(msg);
        String phoneNbr = databaseService.getPhoneNbrFromUsername(getUsername());
        System.out.println("Sending " + msg + " by WhatsApp on " + phoneNbr);
    }
}

public static void main(String[] args) {
    INotifier notifier = new FacebookDecorator(
        new WhatsAppDecorator(
            new Notifier("Geekific")
        )
    );
    notifier.send("Like and Subscribe!!!");
}

